### PR TITLE
[Feat][RayCluster] Introduce the RayClusterStatus.Conditions field

### PR DESCRIPTION
**Pull request by rueian**
## Why are these changes needed?

This is the first part of the process to refine the Ready/Failed status in RayClusterStatus https://docs.google.com/document/d/1bRL0cZa87eCX6SI7gqthN68CgmHaB6l3-vJuIse-BrY

In this first PR, we introduce the new `RayClusterStatus.Conditions` field, borrowed from the idea of k8s Deployment https://github.com/kubernetes/api/blob/857a946a225f212b64d42c68a7da0dc44837636f/apps/v1/types.go#L532-L542

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests

### CODE REVIEW

File: ray-operator/apis/ray/v1/raycluster_types.go [hunk 1]
```diff
@@ -140,6 +140,14 @@ type RayClusterStatus struct {
 	Head HeadInfo `json:"head,omitempty"`
 	// Reason provides more information about current State
 	Reason string `json:"reason,omitempty"`
+
+	// Represents the latest available observations of a deployment's current state.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []RayClusterCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/raycluster_types.go:149]: Why not use `metav1.Condition` instead? https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition

Comment (rueian) [reply ray-operator/apis/ray/v1/raycluster_types.go]: I missed that. Thank you for reminding me.

File: ray-operator/apis/ray/v1/raycluster_types.go [hunk 2]
```diff
@@ -140,6 +140,14 @@ type RayClusterStatus struct {
 	Head HeadInfo `json:"head,omitempty"`
 	// Reason provides more information about current State
 	Reason string `json:"reason,omitempty"`
+
+	// Represents the latest available observations of a deployment's current state.
+	// +patchMergeKey=type
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/raycluster_types.go:145]: Why do we need these?

Comment (astefanutti) [reply ray-operator/apis/ray/v1/raycluster_types.go:145]: This is used by strategic merge patch: https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch.
This is also used by server-side apply, when `listType=map` and `listMapKey` are not specified: https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy.

File: ray-operator/apis/ray/v1/raycluster_types.go [hunk 3]
```diff
@@ -140,6 +140,14 @@ type RayClusterStatus struct {
 	Head HeadInfo `json:"head,omitempty"`
 	// Reason provides more information about current State
 	Reason string `json:"reason,omitempty"`
+
+	// Represents the latest available observations of a deployment's current state.
```
Comment (astefanutti) [inline ray-operator/apis/ray/v1/raycluster_types.go:144]: This line does not seem specific to the conditions field. It can probably be removed.

Comment (rueian) [reply ray-operator/apis/ray/v1/raycluster_types.go]: Updated to `// Represents the latest available observations of a RayCluster's current state.`

File: ray-operator/apis/ray/v1/raycluster_types.go (hunk 2 above)
Comment (rueian) [reply ray-operator/apis/ray/v1/raycluster_types.go:145]: Yes! Thank you @astefanutti for helping me clarify those markers.

File: ray-operator/apis/ray/v1/raycluster_types.go [hunk 4]
```diff
@@ -155,6 +163,14 @@ type RayClusterStatus struct {
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
 }
 
+type RayClusterConditionType string
+
+const (
+	HeadReady RayClusterConditionType = "HeadReady"
```
Comment (andrewsykim) [inline ray-operator/apis/ray/v1/raycluster_types.go:169]: nit: Add a comment for HeadReady

Comment (rueian) [reply ray-operator/apis/ray/v1/raycluster_types.go:170]: Added.

**Comment by andrewsykim** [comment]
@rueian @kevin85421 can we have the API change and implementation in the same PR if possible? Otherwise, if we release v1.2 before the implementation we'd have to revert this PR.

**Comment by kevin85421** [comment]
> can we have the API change and implementation in the same PR if possible? Otherwise, if we release v1.2 before the implementation we'd have to revert this PR.

Do you mean adding `HeadReady` and `RayClusterReplicaFailure` only when we implement them?

**Comment by andrewsykim** [comment]
> Do you mean adding HeadReady and RayClusterReplicaFailure only when we implement them?

Yes, and also the addition of the new `conditions` field in status. We should avoid changing the CRD / API until there is an working / tested minimum implementation. What do you think?

**Comment by andrewsykim** [comment]
@kevin85421 on second thought, I think incremental PRs is fine as long as it's feature gated. I added initial feature gate implementation in https://github.com/ray-project/kuberay/pull/2219, what do you think?

**Comment by rueian** [comment]
I am ok with the feature gate or waiting for the implementation of HeadReady and RayClusterReplicaFailure.

> We should avoid changing the CRD / API until there is an working / tested minimum implementation.

Hi @andrewsykim, do we still need to keep the old CRD / API if the gate is disabled? I thought that could only be achieved by introducing a new CRD version.

**Comment by kevin85421** [comment]
@andrewsykim Does #2219 need to be merged before this PR or the order is flexible?

**Comment by kevin85421** [comment]
@rueian I have synchronized with @andrewsykim today. If the feature flag is off, we will still generate conditions for the CRD, but we will not write anything to conditions. This PR doesn't write anything to `conditions`, so this PR is independent from #2219.

**Review by kevin85421 — APPROVED** [review]

**Pull request merged**